SimPEG/regularization/base.py

```diff
@@ -180,12 +352,290 @@ def deriv2(self, m, v=None):
             R(m) = \\mathbf{W^\\top W}
 
         """
-
-        mD = self.mapping.deriv(self._delta_m(m))
+        f_m_deriv = self.f_m_deriv(m)
         if v is None:
-            return mD.T * self.W.T * self.W * mD
+            return f_m_deriv.T * ((self.W.T * self.W) * f_m_deriv)
+
+        return f_m_deriv.T * (self.W.T * (self.W * (f_m_deriv * v)))
+
+    # TODO move these to general validation functions in code_utils
```

Comment:
This TODO line could be called to be completed and remove it? If not, would be better to remove it and open an issue for it.